File: training_and_examples/scripts/talker_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node

from rclpy.action import ActionClient
from action_msgs.msg import GoalStatus


from example_infs.msg import Uhuy
from example_infs.srv import Print
from example_infs.action import Bruh
from std_msgs.msg import String


class TalkerClientPython(Node):

  def __init__(self):
    super().__init__('talker_client_python')

    self.old_print = ''
    self.old_bruh = 0

    self.get_logger().info('Initiating talker and client...')
    
    # publisher
    self.pub_message = self.create_publisher(String, 'example_msg/message', 10)
    self.pub_uhuy = self.create_publisher(Uhuy, 'example_msg/uhuy', 10)

    # timer
    timer_period = 0.5  # seconds
    self.timer = self.create_timer(timer_period, self.timer_callback)

    # service client
    self.cli_print = self.create_client(Print, 'example_srv/print')

    # action client
    self.cli_act_bruh = ActionClient(self, Bruh, 'example_act/bruh')

    # parameters
    self.declare_parameter('example_param/input', 'STOP')
    self.declare_parameter('example_param/bruh', 0)

    self.service_check()

  def timer_callback(self):
    param_input = self.get_parameter('example_param/input').get_parameter_value().string_value
    count_bruh = self.get_parameter('example_param/bruh').get_parameter_value().integer_value

    msg = String()
    msg.data = 'Halo dunia!'
    self.pub_message.publish(msg)

    uhuy_msg = Uhuy()
    uhuy_msg.uhuy = 'UHUY'
    uhuy_msg.command = 'PRINT'
    self.pub_uhuy.publish(uhuy_msg)

    self.request_print(param_input)
    self.send_goal(count_bruh)

  def service_check(self):
    while ( (not self.cli_print.wait_for_service(timeout_sec=1.0)) and (not self.cli_act_bruh.wait_for_server(timeout_sec=1.0)) ):
      if not rclpy.ok():
        self.get_logger().error('Interrupted while waiting for server. Exiting')
        return

      self.get_logger().warn('Service/Action not available, waiting...')


  def request_print(self, input):
    new_print = input

    if (new_print != self.old_print):
      request = Print.Request()
      request.command = input

      future = self.cli_print.call_async(request)

      def callback_fut_print(_future):
        result = _future.result()
        self.get_logger().info('Service call responded with ' + str(result.success))
        self.old_print = input

      future.add_done_callback(callback_fut_print)    

  # ACTION FUNCTIONS
  def send_goal(self, _count_bruh):

    if ( (self.old_bruh != _count_bruh) and (_count_bruh > 0) ):
      goal_msg = Bruh.Goal()
      goal_msg.count = _count_bruh

      self.get_logger().info('Sending goal')

      goal_future = self.cli_act_bruh.send_goal_async(
        goal_msg,
        feedback_callback=self.feedback_callback)

      # ACTION GOAL RESPONSE CALLBACK
      def goal_response_callback(_future_goal):
        goal_handle = _future_goal.result()
        if not goal_handle.accepted:
          self.get_logger().info('Goal rejected :(')
        else:
          self.get_logger().info('Goal accepted :)')
          
          result_future = goal_handle.get_result_async()

          # ACTION RESULT CALLBACK
          def get_result_callback(_future_result):
            result = _future_result.result().result
            status = _future_result.result().status

            if status == GoalStatus.STATUS_SUCCEEDED:
              self.get_logger().info('Service call responded with ' + str(result.status))
            elif status == GoalStatus.STATUS_ABORTED:
              self.get_logger().info('Goal was aborted')
            elif status == GoalStatus.STATUS_CANCELED:
              self.get_logger().info('Goal was canceled')
            else:
              self.get_logger().info('Unknown result code')

          result_future.add_done_callback(get_result_callback)

        self.old_bruh = _count_bruh

      goal_future.add_done_callback(goal_response_callback)  
      

  # ACTION CALLBACKS
  def feedback_callback(self, feedback):
    self.get_logger().info(str(feedback.feedback.bruh))

def main(args=None):
  rclpy.init(args=args)

  talker_client_python = TalkerClientPython()

  rclpy.spin(talker_client_python)

  # Destroy the node explicitly
  # (optional - otherwise it will be done automatically
  # when the garbage collector destroys the node object)
  talker_client_python.destroy_node()
  rclpy.shutdown()


if __name__ == '__main__':
  main()

File: example_programs/src/talker_client.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>
#include <cstdlib>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"

#include "std_msgs/msg/string.hpp"
#include "example_infs/srv/print.hpp"
#include "example_infs/msg/uhuy.hpp"
#include "example_infs/action/bruh.hpp"

using namespace std::chrono_literals;

using std::placeholders::_1;
using std::placeholders::_2;

using Bruh = example_infs::action::Bruh;
using GoalHandleBruh = rclcpp_action::ClientGoalHandle<Bruh>;

class TalkerClientCpp : public rclcpp::Node {

  public:
  
    // CONSTRUCTOR //
    TalkerClientCpp() : Node("talker_client_cpp") {

      RCLCPP_INFO(this->get_logger(), "Initiating talker and client...");
      // publisher
      pub_message = this->create_publisher<std_msgs::msg::String>("example_msg/message", 10);

      pub_uhuy = this->create_publisher<example_infs::msg::Uhuy>("example_msg/uhuy", 10);

      // timer
      timer = this->create_wall_timer(500ms, std::bind(&TalkerClientCpp::timer_loop, this));

      // service client
      cli_print = this->create_client<example_infs::srv::Print>("example_srv/print");

      // action client
      cli_act_bruh = rclcpp_action::create_client<Bruh>(this, "example_act/bruh");

      // parameters
      this->declare_parameter<std::string>("example_param/input", "bruh");

      this->declare_parameter<int>("example_param/bruh", 0);

      TalkerClientCpp::service_check(); // check if service server online

    }

  private:
    
    std::string old_print, param_input;

    int old_bruh, count_bruh;

    // create timer variable
    rclcpp::TimerBase::SharedPtr timer;

    // create publisher variable
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr pub_message;

    rclcpp::Publisher<example_infs::msg::Uhuy>::SharedPtr pub_uhuy;

    // create service client variable
    rclcpp::Client<example_infs::srv::Print>::SharedPtr cli_print;

    // create action client variable
    rclcpp_action::Client<Bruh>::SharedPtr cli_act_bruh;

    // basically void loop in arduino, but more e p i c
    void timer_loop() {
      this->get_parameter("example_param/input", param_input);
      this->get_parameter("example_param/bruh", count_bruh);

      auto msg = std_msgs::msg::String();
      msg.data = "Halo dunia!";
      pub_message->publish(msg);

      auto uhuy_msg = example_infs::msg::Uhuy();
      uhuy_msg.uhuy = "UHUY";
      uhuy_msg.command = "PRINT";
      pub_uhuy->publish(uhuy_msg);

      request_print(param_input);
      send_goal(count_bruh);
    }

    // make sure all service online (in the beginning only, cause wait_for_service will mess granularity)
    void service_check() {
      while ( (!cli_print->wait_for_service(1s)) && (!cli_act_bruh->wait_for_action_server(1s)) ) {
        if (!rclcpp::ok()) {
          RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for server. Exiting");
          return;
        }
        RCLCPP_WARN(this->get_logger(), "Service/Action not available, waiting...");
      }
    }
 
    void request_print(const std::string& input) {

      if (input != old_print) {

        auto request = std::make_shared<example_infs::srv::Print::Request>();
        request->command = input; 

        cli_print->async_send_request(request, std::bind(&TalkerClientCpp::callback_fut_print, this, _1));  // with a callback function
        old_print = input;
      } 
 
    }

    void callback_fut_print(const rclcpp::Client<example_infs::srv::Print>::SharedFuture future) {
      std::shared_ptr<example_infs::srv::Print::Response> result = future.get();
      RCLCPP_INFO(this->get_logger(), "Service call responded with %s", result->success ? "true" : "false");

      // this function could return the other than void type. might be good for future reference
      // return true;
      // return false if something goes wrong
    }

    // ACTION FUNCTIONS
    void send_goal(const int& _count_bruh) {

      if ((old_bruh != _count_bruh) && (_count_bruh > 0)) {
        auto goal_msg = Bruh::Goal();
        goal_msg.count = _count_bruh;

        RCLCPP_INFO(this->get_logger(), "Sending goal");

        auto send_goal_options = rclcpp_action::Client<Bruh>::SendGoalOptions();
        send_goal_options.goal_response_callback = std::bind(&TalkerClientCpp::goal_response_callback_bruh, this, _1);
        send_goal_options.feedback_callback = std::bind(&TalkerClientCpp::feedback_callback_bruh, this, _1, _2);
        send_goal_options.result_callback = std::bind(&TalkerClientCpp::result_callback_bruh, this, _1);
        
        cli_act_bruh->async_send_goal(goal_msg, send_goal_options);
        old_bruh = _count_bruh;
      }
      
    }

    // ACTION CALLBACKS
    void goal_response_callback_bruh(std::shared_future<GoalHandleBruh::SharedPtr> future) {
      auto goal_handle = future.get();
      if (!goal_handle) {
        RCLCPP_ERROR(this->get_logger(), "Goal was rejected by server");
      } 
      else {
        RCLCPP_INFO(this->get_logger(), "Goal accepted by server, waiting for result");
      }
    }

    void feedback_callback_bruh(
      GoalHandleBruh::SharedPtr,
      const std::shared_ptr<const Bruh::Feedback> feedback) {

      RCLCPP_INFO(this->get_logger(), feedback->bruh.c_str());
    }

    void result_callback_bruh(const GoalHandleBruh::WrappedResult& result) {
      switch (result.code) {
        case rclcpp_action::ResultCode::SUCCEEDED:
          RCLCPP_INFO(this->get_logger(), "Service call responded with %s", result.result->status ? "true" : "false");
          return;
        case rclcpp_action::ResultCode::ABORTED:
          RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
          return;
        case rclcpp_action::ResultCode::CANCELED:
          RCLCPP_ERROR(this->get_logger(), "Goal was canceled");
          return;
        default:
          RCLCPP_ERROR(this->get_logger(), "Unknown result code");
          return;
      }
    }

};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TalkerClientCpp>());
  rclcpp::shutdown();
  return 0;
}
